presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java

```diff
@@ -195,14 +196,70 @@ public static Slice reverse(@SqlType("varchar(x)") Slice slice)
     @SqlType(StandardTypes.BIGINT)
     public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring)
     {
+        return stringPosition(string, substring, 1);
+    }
+
+    @Description("returns index of n-th occurrence of a substring (or 0 if not found)")
+    @ScalarFunction("strpos")
+    @LiteralParameters({"x", "y"})
+    @SqlType(StandardTypes.BIGINT)
+    public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring, @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        if (instance < 0) {
+            return stringPositionFromEnd(string, substring, abs(instance));
+        }
+        return stringPositionFromStart(string, substring, instance);
+    }
+
+    private static long stringPositionFromStart(Slice string, Slice substring, long instance)
+    {
+        if (instance <= 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "'instance' must be a positive or negative number.");
+        }
         if (substring.length() == 0) {
             return 1;
         }
 
-        int index = string.indexOf(substring);
-        if (index < 0) {
-            return 0;
+        int foundInstances = 0;
+        // set the initial index just before the start of the string
+        // this is to allow for the initial index increment
+        int index = -1;
+        do {
+            // step forwards through string
+            index = string.indexOf(substring, index + 1);
+            if (index < 0) {
+                return 0;
+            }
+            foundInstances++;
         }
+        while (foundInstances < instance);
+
+        return countCodePoints(string, 0, index) + 1;
+    }
+
+    private static long stringPositionFromEnd(Slice string, Slice substring, long instance)
+    {
+        if (instance <= 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "'instance' must be a positive or negative number.");
+        }
+        if (substring.length() == 0) {
+            return 1;
+        }
+
+        int foundInstances = 0;
+        // set the initial index just after the end of the string
+        // this is to allow for the initial index decrement
+        int index = string.length();
+        do {
+            // step backwards through string
+            index = string.toStringUtf8().lastIndexOf(substring.toStringUtf8(), index - 1);
```

Comment:
This will get the last index in terms of String characters. But we use this index in `countCodePoints(string, 0, index) + 1`, which is based on bytes.

Reply:
Thanks! I didn't catch it since I skipped the tests with non-ASCII unicode characters